lib/diff.mli

```diff
@@ -1,100 +1,82 @@
-type ('item, 'diff) t = Added of 'item | Removed of 'item | Modified of 'diff
+type ('item, 'diff) entry =
+  | Added of 'item
+  | Removed of 'item
+  | Modified of 'diff
+
 type 'a atomic_modification = { reference : 'a; current : 'a }
+type 'item atomic_entry = ('item, 'item atomic_modification) entry
 
-type 'item atomic_t = ('item, 'item atomic_modification) t
+type ('item, 'diff) map = {
+  same : 'item String_map.t;
+  different : ('item, 'diff) entry String_map.t;
+}
 
-type ('same, 'different) maybe_changed =
-  | Same of 'same
-  | Different of 'different
+type ('same, 'change) maybe_changed = Same of 'same | Changed of 'change
 
 type 'same atomic_maybe_changed =
   ('same, 'same atomic_modification) maybe_changed
 
-type ('same, 'diff) option_ =
-  ('same option, ('same, 'diff) t) maybe_changed
+type 'same atomic_maybe_changed_entry =
+  ('same, 'same atomic_entry) maybe_changed
 
-type 'same atomic_option =
-  ('same, 'same atomic_modification) option_
+type ('same, 'diff) option_ = ('same option, ('same, 'diff) entry) maybe_changed
+type 'same atomic_option = ('same, 'same atomic_modification) option_
 
 type ('same, 'diff) atomic_variant =
   ('same, ('same atomic_maybe_changed, 'diff) maybe_changed) maybe_changed
 
 type ('same, 'diff) list_ = ('same list, 'diff list) maybe_changed
```

Comment:
I think the `list_` should include more than this: it's a list of things than may or may not have changed, and if they changed, it can be an addition, removal or modification.

I believe this will allow to remove the confusing `atomic_maybe_changed_entry` (which is hard to understand what it's for from the name) and hopefully simplify the usage of `list_`.

Reply:
I agree. But that's not always the case with a list diff type. For example with type parameters, it's either same or changed, and the change could be either an added or removed type parameter (we don't have modified here). so I think we have to specify the kind of diff we have as a parameter for type `list_`.

I was wondering if it would make up for a good use case of extensible variants here ?
Like having the top entry type be an extensible type as this: `type ('item, 'diff) entry = ..`
and then extending it as needed with different constructors.
Like this for type paramters: 
```ml 
type ('item, `diff) entry += Added_tp
type ('item, 'diff) entry += Removed_tp
```